cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -165,71 +172,110 @@ func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+			}
+			mig, found := gc.getMig(migRef)
+			if !found {
+				return nil, fmt.Errorf("instance %+v does not belong to any configured MIG", instanceRef)
 			}
-			return nil, fmt.Errorf("instance %+v does not belong to any configured MIG", *instance)
+			return mig.Config, nil
 		}
 	}
 	// Instance doesn't belong to any configured mig.
 	return nil, nil
 }
 
-// RegenerateInstancesCache triggers instances cache regeneration under lock.
-func (gc *GceCache) RegenerateInstancesCache() error {
+func (gc *GceCache) getMig(migRef GceRef) (*MigInformation, bool) {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+	mig, found := gc.migs[migRef]
+	return mig, found
+}
+
+// RegenerateInstanceCacheForMig triggers instances cache regeneration for single MIG under lock.
+func (gc *GceCache) RegenerateInstanceCacheForMig(migRef GceRef) error {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
-
-	return gc.regenerateCache()
+	return gc.regenerateInstanceCacheForMigNoLock(migRef)
 }
 
-// internal method - should only be called after locking on cacheMutex.
-func (gc *GceCache) regenerateCache() error {
-	newInstancesCache := make(map[GceRef]Mig)
+func (gc *GceCache) regenerateInstanceCacheForMigNoLock(migRef GceRef) error {
+	klog.V(4).Infof("Regenerating MIG information for %s", migRef.String())
+
+	// Here we are on purpose not deleting old entries pointing to migRef from gc.instanceRefToMig.
+	// They are no harm as same GceRef should not be used for new instance and instance should not move from one Mig to
+	// the other.
+	// And garbage-collection will be enforces as whole cache is invalidated hourly.
 
-	for _, migInfo := range gc.GetMigs() {
-		mig := migInfo.Config
-		klog.V(4).Infof("Regenerating MIG information for %s", mig.GceRef().String())
+	basename, err := gc.GceService.FetchMigBasename(migRef)
+	if err != nil {
+		return err
+	}
+	gc.updateMigBasename(migRef, basename)
 
-		basename, err := gc.GceService.FetchMigBasename(mig.GceRef())
+	instances, err := gc.GceService.FetchMigInstances(migRef)
+	if err != nil {
+		klog.V(4).Infof("Failed MIG info request for %s: %v", migRef.String(), err)
+		return err
+	}
+	for _, instance := range instances {
+		instanceRef, err := GceRefFromProviderId(instance.Id)
 		if err != nil {
 			return err
 		}
-		gc.updateMigBasename(mig.GceRef(), basename)
+		gc.instanceRefToMigRef[instanceRef] = migRef
+	}
+	return nil
+}
+
+// RegenerateInstancesCache triggers instances cache regeneration under lock.
+func (gc *GceCache) RegenerateInstancesCache() error {
+	gc.cacheMutex.Lock()
+	defer gc.cacheMutex.Unlock()
+
+	return gc.regenerateInstancesCacheNoLock()
+}
 
-		instances, err := gc.GceService.FetchMigInstances(mig.GceRef())
+func (gc *GceCache) regenerateInstancesCacheNoLock() error {
+	gc.invalidateInstanceCacheNoLock()
+	for _, migRef := range gc.getMigRefs() {
+		err := gc.regenerateInstanceCacheForMigNoLock(migRef)
 		if err != nil {
```

Comment:
nit: if err := ...; err != nil { return err }

Reply:
less readable to me